### PR TITLE
GIX-2080: Migrate e2e accounts.spec to tokens page

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Enable my tokens flag for mainnet.

In this PR, set the flag to true for the accounts.spec e2e test and change the expects and interactions to adapt to the new UI.

# Changes

* Set flag `ENABLE_MY_TOKENS` to `true` in accounts.spec and adapt it to the new UI.
* Add convenient methods to different POs to adapt the e2e test to the new UI.

# Tests

Only test changes.

# Todos

- [ ] Add entry to changelog (if necessary).

### CODE REVIEW

**Comment by lmuntaner** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #4183** <a href="https://app.graphite.dev/github/pr/dfinity/nns-dapp/4183?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dfinity/nns-dapp/4183?utm_source=stack-comment).

File: frontend/src/tests/page-objects/TokensTable.page-object.ts [hunk 1]
```diff
@@ -44,6 +45,21 @@ export class TokensTablePo extends BasePageObject {
     return undefined;
   }
 
+  async waitForRowByName(projectName: string): Promise<void> {
```
Comment (lmuntaner) [inline frontend/src/tests/page-objects/TokensTable.page-object.ts:48]: This was needed because the table takes some time to load the data and I needed a way to way for the presence of a specific row.

File: frontend/src/tests/utils/utils.test-utils.ts [hunk 2]
```diff
@@ -44,3 +44,8 @@ export const isNotVisible = (element: Element): boolean => {
     return nonNullish(heightString) && Number(heightString) === 0;
   }
 };
+
+export const waitForMilliseconds = (milliseconds: number): Promise<void> =>
```
Comment (lmuntaner) [inline frontend/src/tests/utils/utils.test-utils.ts:48]: We have this util for production, but it was not usable within the e2e tests. Therefore, I copied within `tests` directory.

**Comment by lmuntaner** [comment]
@dskloetd please review

File: frontend/src/tests/page-objects/TokensTable.page-object.ts [hunk 3]
```diff
@@ -44,6 +45,21 @@ export class TokensTablePo extends BasePageObject {
...
+    let row = await this.findRowByName(projectName);
+    let counter = 1;
+    const MAX_TRIES = 10;
+    const WAIT_MILLISECONDS = 1_000;
+    while (isNullish(row) && counter < MAX_TRIES) {
+      await waitForMilliseconds(WAIT_MILLISECONDS * counter);
```
Comment (dskloetd) [inline frontend/src/tests/page-objects/TokensTable.page-object.ts:54]: I think it would be cleaner to use the existing `querySelectorCount` or `countByTestId` and wait for rows one by one until you have one that matches.

Comment (lmuntaner) [reply frontend/src/tests/page-objects/TokensTable.page-object.ts:54]: Do you mean instead of `findRowByName`?

How would the count help me find the specific row?

Do you mean to create a `waitForRowsCount(amount)`?

Another idea I had was to use the projectName in the testId of the row. In that case I could select it directly and use `waitFor`.

Comment (dskloetd) [reply frontend/src/tests/page-objects/TokensTable.page-object.ts:54]: I mean to avoid explicit `waitForMilliseconds`.
Something like
```
for (let i = 1;; i++) {
  const row = TokensTableRowPo.countUnder({ element: this.root, count: i}).at(-1);
  if (await row.getProjectName() === projectName) {
    return row;
  }
}
```
Implementation of `TokensTableRowPo.countUnder` would be similar to `SelectUniverseCardPo.countUnder`.

Putting the token symbol in the test ID is also fine with me but it might interfere with an existing test ID.

Comment (lmuntaner) [reply frontend/src/tests/page-objects/TokensTable.page-object.ts]: I think it's nicer to use another attribute for the project name: `data-title`.

I implemented it and it looks cleaner IMO.

Let me know what you think.

File: frontend/src/tests/page-objects/TokensTableRow.page-object.ts [hunk 4]
```diff
@@ -17,6 +17,18 @@ export class TokensTableRowPo extends BasePageObject {
     );
   }
 
+  static async byTitle({
```
Comment (lmuntaner) [inline frontend/src/tests/page-objects/TokensTableRow.page-object.ts:20]: @dskloetd I added this static method to the row.

File: frontend/src/tests/page-objects/TokensTable.page-object.ts [hunk 5]
```diff
@@ -34,14 +34,7 @@ export class TokensTablePo extends BasePageObject {
...
   ): Promise<TokensTableRowPo | undefined> {
-    const rows = await this.getRows();
-    for (const row of rows) {
-      const name = await row.getProjectName();
-      if (name === projectName) {
-        return row;
-      }
-    }
-    return undefined;
+    return TokensTableRowPo.byTitle({ element: this.root, title: projectName });
```
Comment (lmuntaner) [inline frontend/src/tests/page-objects/TokensTable.page-object.ts:37]: Just select directly instead of looping.

**Comment by lmuntaner** [comment]
@dskloetd ready for another review

File: frontend/src/tests/page-objects/TokensTableRow.page-object.ts [hunk 6]
```diff
@@ -17,6 +17,18 @@ export class TokensTableRowPo extends BasePageObject {
...
+    element,
+    title,
+  }: {
+    element: PageObjectElement;
+    title: string;
+  }): Promise<TokensTableRowPo> {
```
Comment (dskloetd) [inline frontend/src/tests/page-objects/TokensTableRow.page-object.ts:26]: This doesn't need to be async, does it?
Same for the caller and the caller of the caller.

File: frontend/src/tests/e2e/accounts.spec.ts [hunk 7]
```diff
@@ -14,55 +20,70 @@ test("Test accounts requirements", async ({ page, context }) => {
...
+    (await tokensTablePo.getRowsData()).map(({ projectName }) => projectName);
   expect(await accountNames()).toEqual([mainAccountName, subAccountName]);
 
   // The linked account should still be present after refresh
   await page.reload({ waitUntil: "load" });
+
+  // We wait until the table is loaded.
+  await tokensTablePo.waitFor();
+  // We wait until the subaccount row is loaded.
+  const subaccountRow = await tokensTablePo.findRowByName(subAccountName);
```
Comment (dskloetd) [inline frontend/src/tests/e2e/accounts.spec.ts:50]: I think this doesn't need `await` anymore.

File: frontend/src/tests/page-objects/TokensTable.page-object.ts [hunk 8]
```diff
@@ -34,14 +34,7 @@ export class TokensTablePo extends BasePageObject {
   async findRowByName(
```
Comment (dskloetd) [inline frontend/src/tests/page-objects/TokensTable.page-object.ts:34]: Does `find` still make sense with the current implementation or should it be `get`?
This will return something regardless of whether row is there yet so "find" seems inaccurate.

File: frontend/src/tests/e2e/accounts.spec.ts [hunk 9]
```diff
@@ -72,6 +93,5 @@ test("Test accounts requirements", async ({ page, context }) => {
   expect(transactions).toHaveLength(1);
   const transaction = transactions[0];
 
-  expect(await transaction.getIdentifier()).toBe(`From: ${mainAccountAddress}`);
```
Comment (dskloetd) [inline frontend/src/tests/e2e/accounts.spec.ts:75]: Why is this removed?

File: frontend/src/tests/e2e/accounts.spec.ts [hunk 10]
```diff
@@ -14,55 +20,70 @@ test("Test accounts requirements", async ({ page, context }) => {
...
+  expect(await mainAccountRow.getBalance()).toEqual("20.00 ICP");
+  expect(await subaccountRow.getBalance()).toEqual("0 ICP");
 
-  const subAccountAddress =
-    await nnsAccountsPo.getAccountAddress(subAccountName);
-  await nnsAccountsPo.getMainAccountCardPo().click();
+  await mainAccountRow.click();
   await appPo.getWalletPo().getNnsWalletPo().transferToAccount({
     accountName: subAccountName,
-    expectedAccountAddress: subAccountAddress,
```
Comment (dskloetd) [inline frontend/src/tests/e2e/accounts.spec.ts:49]: Is there no way to see the account address once we launch the my tokens table? Isn't it visible when you click on the receive icon?

Comment (lmuntaner) [reply frontend/src/tests/e2e/accounts.spec.ts:49]: Yes, it's there.

But is it necessary for this test?

Maybe we can add another e2e test, no? Or is there a reason to have such a test here?

File: frontend/src/tests/e2e/accounts.spec.ts (hunk 9 above)
Comment (lmuntaner) [reply frontend/src/tests/e2e/accounts.spec.ts:75]: Because we don't have the address in the Tokens table.

File: frontend/src/tests/page-objects/TokensTableRow.page-object.ts (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/tests/page-objects/TokensTableRow.page-object.ts]: True, removed.

File: frontend/src/tests/page-objects/TokensTable.page-object.ts (hunk 8 above)
Comment (lmuntaner) [reply frontend/src/tests/page-objects/TokensTable.page-object.ts]: done

File: frontend/src/tests/e2e/accounts.spec.ts (hunk 9 above)
Comment (lmuntaner) [reply frontend/src/tests/e2e/accounts.spec.ts]: reverted

File: frontend/src/tests/e2e/accounts.spec.ts (hunk 10 above)
Comment (lmuntaner) [reply frontend/src/tests/e2e/accounts.spec.ts:49]: Done

**Comment by lmuntaner** [comment]
@dskloetd ready for another review

File: frontend/src/tests/page-objects/Accounts.page-object.ts [hunk 11]
```diff
@@ -74,4 +74,16 @@ export class AccountsPo extends BasePageObject {
...
+
+  async getAccountAddress(accountName: string): Promise<string> {
+    const row = this.getNnsAccountsPo()
+      .getTokensTablePo()
+      .getRowByName(accountName);
+    await row.clickReceive();
+    const receiveModalPo = this.getReceiveModalPo();
+    await receiveModalPo.waitFor();
+    const address = await receiveModalPo.getAddress();
+    await receiveModalPo.clickFinish();
```
Comment (dskloetd) [inline frontend/src/tests/page-objects/Accounts.page-object.ts:86]: "Finish" does an extra unnecessary reload of the transactions, right?
We could use "close" instead to avoid that.

**Review by dskloetd — APPROVED** [review]

File: frontend/src/tests/page-objects/Accounts.page-object.ts (hunk 11 above)
Comment (lmuntaner) [reply frontend/src/tests/page-objects/Accounts.page-object.ts]: done
